### ISSUE TITLE
FlutterNativeView.runFromBundle should support parameters

### ISSUE THREAD

**Issue by mehmetf** [flutter/flutter#15798]
This is important for projects who want to integrate Flutter into their existing apps. There's currently no reasonable way to treat `FlutterView` like a regular Android view:

- It does not support different endpoints in Dart. You can specify different endpoints via runFromBundle but if you expose multiple entry points in Dart, the AOT compiler tree-shakes them.
- There's no easy way of passing parameters to Dart the way you would pass parameters between Android activities via an Intent. The expected way right now is for the Dart app to run and register callbacks that are tied to a PlatformChannel and then the Android shell calls that channel. This works fine for standalone Flutter apps but for a small Flutter activity integrated into a larger Android app, it is a pain.

If we support parameters in runFromBundle, we could do something like this on Dart:

```dart
void main([List<String> args]) {
  if (args?.length != 1) return;
  switch (args[1]) {
    case 'page1':
      runApp(new Page1());
      break;
    case 'page2':
      runApp(new Page2());
      break;
  }
}
```

> Ideally the params can be passed further down the app and the app can decide how to route pages but you get the idea.

I assume the same thing will be a problem in iOS as well.

**Comment by chinmaygarde** [comment]
On iOS, we could do this by specifying the launch arguments as an array (or maybe a map) on [`FlutterDartProject`](https://github.com/flutter/engine/blob/2114a88b5ee0c5240f2aec6cde164798d00b61b4/shell/platform/darwin/ios/framework/Headers/FlutterDartProject.h#L13). Dart IO seems to also populate [`Platform.environment`](https://api.dartlang.org/stable/1.24.3/dart-io/Platform/environment.html). Maybe putting these parameters in that map is more idiomatic.

**Comment by mehmetf** [comment]
Heart has to work around this issue in their prototype at the moment which requires them to pass data into FlutterView from their native activity at construction time. For now the workaround is to construct the view and have Flutter ask for the data via PlatformChannel once it is ready.

**Comment by Hixie** [comment]
Does setInitialRoute address this suitably for you? It's how the design currently intends for you to pass data from the OEM side to the Dart before the Dart code starts up.

**Comment by mehmetf** [comment]
@Hixie That indeed works. I think it should be enough to cover this use case. It would be nice to be able to pass structured data but definitely low priority.

**Comment by Hixie** [comment]
Sending structured data is indeed something that would be interesting, but we'd rather not reinvent system channels just for the initial route, so our theory is that if you want to send structured data, you'd send a sentinel that indicates to the dart code "hey, contact the OEM side to find out what the structured data is".

**Comment by mehmetf** [comment]
Thanks @Hixie. That is precisely the workaround I had recommended.

Just as an anecdote for people stumbling on this: `initialRoute` only works if you are in foreground and have an active `FlutterView`. It does not work if you are trying to call an arbitrary Dart fn in the background via `FlutterNativeView`. For instance, the [AlarmManager plugin](https://github.com/flutter/plugins/blob/4af10ced41ca3cd2249463d569abc1d13548cba5/packages/android_alarm_manager/example/lib/main.dart#L56) currently supports VoidCallback type methods to be run.

I imagine we will eventually need to pave the path to setting up PlatformChannel's in the background. These channels already allow you to pass parameters and return values from either side. That's ultimately what AlarmManager plugin needs to be able to do. Since we have no customer use case demanding this (yet), I am OK closing it.

**Comment by pinkfish** [comment]
The use case for this is firebase messaging isn't it?  Right now firebase messaging drops messages entirely on the floor for android.  It should really be able to call into the native view, and pass in the remote message, to do some useful background processing.  This makes firebase messaging a lot less useful.
